Tables rendering: fixes and improvements

poire-z opened the pull request:
It started with the 30 lines first commit, to support symbol fonts (details in https://github.com/koreader/koreader/issues/4196#issuecomment-435587002), and...:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49339039-b5449f00-f62b-11e8-8613-0d98631b0d70.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/49339005-fdaf8d00-f62a-11e8-8ce4-351ebefbf578.png)</kbd>

and it ended up with 1500+ lines so we can get instead:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49339042-d1484080-f62b-11e8-910f-c6adc19558c5.png)</kbd>

<kbd>![new_symbol2](https://user-images.githubusercontent.com/24273478/49339010-220b6980-f62b-11e8-87b0-973892dbb8c2.png)</kbd>

Fixed all the buggy stuff I met in that journey (including a right border drawing bug, noticable in all the screenshots in #233: the right border was thicker - and now the various `Vocalu-lo` are no more cut - and the background issue noticed in https://github.com/koreader/crengine/pull/233#issuecomment-441317099).

Main improvements and changes:

More adequate computation of column widths, so they don't go smaller than the longest word inside a cell. This should avoid words being cut, and a proper sizing when a cell content uses a different font than the cell itself.
Added `getRenderedWidths()` to get the min and max content width from any node. Used for now for sizing table cells, but could be used for other things (like floats).
Added support for table and cells width when specified with CSS (there was support for that when specified as HTML tag attributes, which worked only with plain HTML documents, not with EPUBs).
When no table width specified, defaults to "shrink to fit" (as most browsers do) and no more to 100%.
Implement `vertical-align: middle/bottom` for table cells.  Closes https://github.com/koreader/koreader/issues/2846
Details in the individual commit messages.

More comparisons (stuff look constrained in the After screenshots, because I used `padding:0` for easier testing - we'd probably find some padding in regular books):

Before:
<kbd>![old1](https://user-images.githubusercontent.com/24273478/49339120-4cf6bd00-f62d-11e8-9121-8637900741ca.png)</kbd>
After:
<kbd>![new1](https://user-images.githubusercontent.com/24273478/49339121-4ff1ad80-f62d-11e8-9162-a050569b8f4c.png)</kbd>

Before:
<kbd>![old2](https://user-images.githubusercontent.com/24273478/49339124-57b15200-f62d-11e8-9ab4-322e7ed05539.png)</kbd>
After:
<kbd>![new2](https://user-images.githubusercontent.com/24273478/49339125-597b1580-f62d-11e8-99c2-34bf06c26797.png)</kbd>

Before:
<kbd>![old3](https://user-images.githubusercontent.com/24273478/49339129-639d1400-f62d-11e8-82dd-3a32e54a5960.png)</kbd>
After:
<kbd>![new3](https://user-images.githubusercontent.com/24273478/49339130-65ff6e00-f62d-11e8-854e-85e744039546.png)</kbd>
(There must be a bug in the font with the first symbol glyph, as its width is different than its rendered width... the other symbols glyphs are ok.)

Some notes and questions:
Supporting cells specified widths is a bit questionable, as I feel it's always better when not, and just using the computed width according to min/max content width feels always good, and specified width can be boggus and change with DPI. We'll see if we need to remove that.

Unless tables are set with a specified width, they will now take only the width necessary to display cells content (shrink to fit) just like current browsers do.
This might be questionable: I feel like always taking 100% for unstyled tables is most often better than not on our devices (otherwise, smaller tables are stuck to the left, which feels strange with justified text... a 100% width table feels like a justified table :)

Got this feeling with most Wikipedia EPUBs, that looked before/after like:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49339210-f0949d00-f62e-11e8-8f2c-c2b29b1f3ca5.png)</kbd> <kbd>![image](https://user-images.githubusercontent.com/24273478/49339215-fc805f00-f62e-11e8-8a87-c316378f3a59.png)</kbd>
(there is actually a `style="width:9em"` on the first TD in a row, that's the reason they look aligned)

What do you think?
If we'd want to restore previous behaviour for unstyled tables, we could just add to epub.css: `table { width: 100%; }`.

I'll add a few Style tweaks so people can choose alternatives (ignore table and td widths, make tables full width, make tables shrink, center shrinked tables...) but we should choose the most adequate default.

This is a followup to #233, and change quite a bit of what @frankyifei introduced in #35 and #38.

There might be something left to do for CJK, but I can't judge.
I left some code related to CJK commented out in `getRenderedWidths()`, as logically, one CJK char being itself a word and candidate for wrap, we could get when sizing (in this example, there's a witdh=100% specified in the middle TD...)
<kbd>![image](https://user-images.githubusercontent.com/24273478/49339309-048cce80-f630-11e8-8e07-d6434354afe3.png)</kbd>
but not doing it gives a better looking:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49339285-d5765d00-f62f-11e8-83c0-091b57927bea.png)</kbd>
@frankyifei or other CJK readers: any feeling about that? Any minimal number of CJK chars to aggregate on a single line, or are there some kind of rules for deciding that?

poire-z commented:
Not to be merged/bumped before 2018.12 is released ! as there will probably be new bugs or glitches... :|

Frenzie commented:
Oh, I forgot all about that. :-P

I'll probably toss in a MuPDF default scaleByDPI thingy (for EPUB) before I do, and also I don't really feel like it right now. :-)

Eduardomb22 commented:
> This might be questionable: I feel like always taking 100% for unstyled tables is most often better than not on our devices (otherwise, smaller tables are stuck to the left, which feels strange with justified text... a 100% width table feels like a justified table :)
I have this feeling too, maybe this should be the default https://user-images.githubusercontent.com/24273478/49339210-f0949d00-f62e-11e8-8f2c-c2b29b1f3ca5.png

 You are a monster. Thanks a lot for commit!!

poire-z commented:
There's also the thing that publishers expect the common rendering on the right (the new one), and most (well, it's the case of 2 of your books I tested with - 1 one of mine did not) will set in their publisher styles `table { width: 100%; }`, so it will look like the one on the left (the old one), as you can see when viewing your books in Calibre: tables take the full width.

Same KOReader Wikipedia EPUB displayed by Calibre:
<kbd>![calibre](https://user-images.githubusercontent.com/24273478/49345864-e56f5a80-f68a-11e8-85e6-b1d0e2de0fbe.png)</kbd>

It's a bit more messy with Wikipedia EPUBs, as tables are set in various different ways (classnames: infobox, infobox_v2, infobox_v3, navbox...), with sometimes the width set on the TABLE, sometimes on some TD, sometimes on a surrounding DIV (where we don't support width)...
So, as KOReader is the publisher of the wikipedia EPUBs :) we could set `table { width: 100%; }` in our wikipedia stylesheet.css. But with `!important` (impossible to get back the ones specified in wikiepdia html) or not (may need a style tweak to get the 100% again)?

I yet have to see how it look on other regular books, tables are not so common :)

So, 3 levels of tweaking:
- our epub.css: either nothing, or `table { width: 100%; }`, or `table { width: auto; margin-left: auto; margin-right: auto }` to have small tables centered.
- wikipedia stylesheet.css (not sure yet what I prefer)
- style tweaks: for a wide choice of personal preferences

So, still thinking :)

Frenzie commented:
> or other CJK readers: any feeling about that? Any minimal number of CJK chars to aggregate on a single line, or are there some kind of rules for deciding that?

Is that a Wikipedia page? I'll just ask my standard question: what do the browsers do? (Because even if there's no written standard, there's often a de facto one.)

poire-z commented:
> Is that a Wikipedia page? I'll just ask my standard question: what do the browsers do?

Yes, some random _mixed chinese/english_ page I once found and now use for testing:
https://zh.wikipedia.org/wiki/Key_(%E9%81%8A%E6%88%B2%E5%93%81%E7%89%8C)

Browsers have a luxury we don't have, they can overflow :) Which is what firefox does:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49382246-3e380500-f716-11e8-86d6-cf4155b0a76b.png)</kbd>
In crengine, we try to keep specified widths and reduce others with unspecified widths (but if none: all columns) to the now-computed min-content-width.
In the 2nd screenshot, the full CJK-sequence-considered-as-a-single-word made that min-content-width, but even there, it was too much, so it was still further reduced and 1 CJK char ended up wrapped.
In the 1st screenshot, I said that a single CJK char make the min-content-width, so crengine went ahead: as it needs 100% for the 2nd cell, it reduced the 1st to the min it can: the longest CJK char width.
This example page/table is a bit twisted, but it gives an extreme sample case.

But OK, testing with just extracting some text from this table and putting it in a simple table with no width specified, Firefox does not mind wrapping down to a single CJK char:

![image](https://user-images.githubusercontent.com/24273478/49383322-bef80080-f718-11e8-9e69-c8bc5b09a33d.png) (right truncated)
![image](https://user-images.githubusercontent.com/24273478/49383343-cd461c80-f718-11e8-91a0-5adc0a7ce437.png) (right truncated)
![image](https://user-images.githubusercontent.com/24273478/49383365-dafba200-f718-11e8-9c52-c515324b9d95.png)
but it seems to additionally try to keep final heights balanced when computing widths. crengine can't easily do that.

Same simpler table with this PR crengine, with the is_cjk stuff still commented out (so, full sequence of CJK considered as a unbreakable unit):
![image](https://user-images.githubusercontent.com/24273478/49383826-f1eec400-f719-11e8-8683-6ae916327fcc.png)

With the is_cjk uncommented (so, each CJK char considered a small unbreakable unit):
![image](https://user-images.githubusercontent.com/24273478/49383916-21053580-f71a-11e8-9c97-d660e3ae23a2.png)

Same with some non-CJK slipped in:
![image](https://user-images.githubusercontent.com/24273478/49383969-46923f00-f71a-11e8-8b99-6a0f5462b9c2.png)

poire-z commented:
No more Travis CI auto-check ?
I'll bump this, and just add `table {width: 100%;}` to wikipedia.lua stylesheet.css, and a few style tweaks, and we'll see the feedback we get.

Frenzie commented:
If that's not in the styles, I indeed think that'd be good to add. It's fairly standard.

Frenzie commented:
Travis is probably just a hiccup.

poire-z commented:
Ok, added.
And fixed a small issue that I just witnessed when <kbd>![image](https://user-images.githubusercontent.com/24273478/49465695-829ed000-f7fe-11e8-98bc-696533137ef9.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/49465611-53885e80-f7fe-11e8-9481-b599e2ce302b.png)</kbd>

Fixed, showing how BLOCKQUOTE is now rendered:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49465819-bf6ac700-f7fe-11e8-9395-90e75b15629f.png)</kbd>

Illustration for the P 1.2em and need for BLOCKQUOTE 2em:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49465937-0f498e00-f7ff-11e8-9cf9-5a21daf70f82.png)</kbd>

poire-z commented:
OK, switched to using margin-left/right instead of padding-left/right as suggested by Frenzie in a comment to the (now gone) epub.css previous commit.

Frenzie commented:
If you want me to read over the actual code, that won't be for today. CSS is just something I do at a glance. ;-)

poire-z commented:
Not forcing that on you :) I'll merge and bump all that tomorrow.

frankyifei commented:
For Chinese, one character is one word and there is no rules for minimal characters in a line. I don't know about Japanese and Korean. Actually it looks much better now after your change. There is a W3C format requirement for Japanese, which is most comprehensive. [Japanese text layout](https://www.w3.org/TR/jlreq/#processing_of_tables)

poire-z commented:
Thanks for the feedback!
So going on and merging.

poire-z commented:
There's definitely a rendering degradation (when some image touch the right side) because of some commits in this PR, as noticed in https://github.com/koreader/koreader/issues/4381#issuecomment-445000816.
Before:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49610910-2ffd1980-f9a0-11e8-867c-6bb3f5164baa.png)</kbd>

After:
<kbd>![image](https://user-images.githubusercontent.com/24273478/49610819-f1675f00-f99f-11e8-8afe-bea9125b73e0.png)</kbd>

Frenzie commented:
I can't speak for Asian typography, but imo it doesn't make much sense to wrap just because you can. Something like the screenshot above clearly illustrates that.

poire-z commented:
It's not so much _wrap just because you can_ than _wrap just because you need to to make more room for the other longer columns/cells and because it's allowed_ :)
In the new algorithm, we compute min widths (was 2 symbol glyphs before - 1 now), and later, when computing the columns width, it sees it need a lot of room for the other columns, so it decreases the 2nd one as needed from its max-width (2 symbol glyphs) by still ensuring min-width, so to something in between its max-width and its min-width (so less than the would-be-nice 2 symbol glyphs width).

Frenzie commented:
That "because it's allowed" is the "just because you can" part, because it clearly isn't needed. In fact the whole thing takes up more space! :-P

poire-z commented:
But you haven't seen the other lines that are not in the screenshot, so may be the whole thing takes less space :) (actually, it doesn't).
But yes, this is a simple extension to the original one-pass algorithm, still keeping it single-pass. I did give some thoughts (when noticing the CJK thing mentionned above) about how to make it account for the resulting heights ... but did not get any simple idea.

poire-z commented:
Any feeling about the BLOCKQUOTE style added some weeks ago?
https://github.com/koreader/crengine/blob/0498331aa2d6b3f893b02631e2a528dda8af3904/cr3gui/data/epub.css#L47-L52

Somehow, the empty space on the right is often bothering me: it makes the blockquoted text too constrained.
Most browsers even use margin-left and margin-right both = 40px (why absolute units ?!).
Only Opera (thus also using px) does not balance them, and has the left=40px but the right=10px
(https://github.com/sw4/revert.css/blob/eae9e817e9fb6a5885504704f69e06e9dd828cd0/revert_opera.css#L49-L52), which looks nicer to me.

I feel using something like

Frenzie commented:
What if you reduce it to `1em` on both sides?

poire-z commented:
There's the thing mentionned above that our `<P>` uses `text-indent: 1.2em` (which I like), and only using 1em on both sides would not make a BLOCKQUOTE stands out in some cases (it could be ambiguous when one of the P or the BLOCKQUOTE is only a single or a few lines).

Frenzie commented:
I'd say no margin on the right isn't the biggest deal but perhaps still 1em on the right to keep it more in line with expected typographic conventions?

virxkane commented:
```diff
diff --git a/crengine/src/lvxml.cpp b/crengine/src/lvxml.cpp
index a4223834..9a0c18bf 100644
--- a/crengine/src/lvxml.cpp
+++ b/crengine/src/lvxml.cpp
@@ -3965,7 +3965,7 @@ static const char * AC_PARAM[] = {"param", "param", NULL};
 static const char * AC_IMG[]= {"img", NULL};
 static const char * AC_TD[] = {"td", "td", "th", NULL};
 static const char * AC_TH[] = {"th", "th", "td", NULL};
-static const char * AC_TR[] = {"tr", "tr", "thead", "tfoot", "tbody", NULL};
+static const char * AC_TR[] = {"tr", "tr", "td", NULL};
 static const char * AC_DIV[] = {"div", "p", NULL};
 static const char * AC_TABLE[] = {"table", "p", NULL};
 static const char * AC_THEAD[] = {"thead", "tr", "thead", "tfoot", "tbody", NULL};
@@ -4016,6 +4016,9 @@ HTML_AUTOCLOSE_TABLE[] = {
     AC_TABLE,
     NULL
 };
+// Note: AC_TD and AC_TR may kill a table nested inside an other table,
+// so such nested tables won't work in a .html file (but will in a .epub
+// that uses the NON-auto-closing ldomDocumentWriter).
 
 
 // base64 decode table
 ```
 
This broke tag autoclosing for ```<tr>``` with this file [autoclose-test2.html.zip](https://github.com/koreader/crengine/files/5039875/autoclose-test2.html.zip)
How it better to fix?

poire-z commented:
> How it better to fix?

I have no idea.
In this PR, it's possible I did that quick change without too much thinking- and I don't remember exactly what this change was fixing.
Since then, I added this comment, which might help understanding how that AC_ business works and deciding the best values:
https://github.com/koreader/crengine/blob/8a15844847f2c0ba264372e61de178472beed8c6/crengine/src/lvxml.cpp#L5866-L5906

So, may be reverting it, if the original makes sense with my comments, is the right thing (no time right now to re-read it and put my mind on it).

But your test HTML files is quite extreme with the absence of lots of closing tags. We might not be able to do magic as what to close or not might be ambiguous and there are multiple options that can all break a thing.

With your test file as-is, with your not-closing DDs, we already got something bad (and I don't know if the issue is with TABLE TR or TD autoclosing, or the DD autoclosing):
<kbd>![image](https://user-images.githubusercontent.com/24273478/89625282-eca32e80-d897-11ea-8d4c-3bd20dc92c75.png)</kbd>

If I remove all your DD, we got something a bit better as the TABLE element works (but I have no idea where the rest of the content is gone :):
<kbd>![image](https://user-images.githubusercontent.com/24273478/89625151-a5b53900-d897-11ea-8198-bf6fa97da422.png)</kbd>

virxkane commented:
> But your test HTML files is quite extreme with the absence of lots of closing tags. We might not be able to do magic as what to close or not might be ambiguous and there are multiple options that can all break a thing.

This is a slightly simplified snippet from a real problematic chm file.
And in old CoolReader (for example 3.2.38) this file rendered more correctly. But sometimes when I change window size table can crumble.

virxkane commented:
Yet another question
```diff
+// Note: AC_TD and AC_TR may kill a table nested inside an other table,
+// so such nested tables won't work in a .html file (but will in a .epub
+// that uses the NON-auto-closing ldomDocumentWriter).
```
This is related to previous line ```static const char * AC_TR[] = {"tr", "tr", "thead", "tfoot", "tbody", NULL};``` or the new line ```static const char * AC_TR[] = {"tr", "tr", "td", NULL};```, or both?

poire-z commented:
I guess I added that comment with the new line `TR[] = {"tr", "tr", "td", NULL};, ` - so I guess it applies to my new line.
The sad thing is that I/we don't know anymore what I solved with that new lines :/ But, knowing how I think and how I makde comments :) I guess that what I solved was more important than the "caveat" I put in that comment :/

poire-z commented:
So, if I trust my comment:
https://github.com/koreader/crengine/blob/8a15844847f2c0ba264372e61de178472beed8c6/crengine/src/lvxml.cpp#L5866-L5906

Current `AC_TR[] = {"tr", "tr", "td", NULL};` : 
When meeting a TR, we check the current node and its parent until we find a TR or TD. if we find one, we autoclose them, and then insert the TR as a sibling of these.

Old `AC_TR[] = {"tr", "tr", "thead", "tfoot", "tbody", NULL};`
When metting a TR, we check the current node and its parent until we find a TR, THEAD, TBODY. If we find one, we autoclose them, and then insert the TR as sibling of these.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/tbody
> The HTML Table Body element (`<tbody>`) encapsulates a set of table rows (`<tr>` elements), indicating that they comprise the body of the table (`<table>`).

So, it feels wrong with the "old rule" to autoclose a TBODY when meeting the TR it contains.
So, I'm not even sure I fixed an issue I really met when changing it: may be it was just the DOM that was looking wrong.

With this sample HTML:
```html
<html lang="">
  <body>
ABC
<table style="border: 1px solid black">
  <THEAD> <TR> <TD>cell1<TD>cell2 </TR> </THEAD>
  <TBODY> <TR> <TD>cell1<TD>cell2 </TR> </TBODY>
</table>
DEF
  </body>
</html>
```

Old rule:
<kbd>![image](https://user-images.githubusercontent.com/24273478/89637012-d7cf9680-d8a9-11ea-8bc5-bce8234a2056.png)</kbd>

Current (my) rule:
<kbd>![image](https://user-images.githubusercontent.com/24273478/89637133-06e60800-d8aa-11ea-8772-6809ba4ff4a1.png)</kbd>

So, the old rule is wrong: it might not change much the table layout to autoclose a TBODY and have its TR direct children of the TABLE - except when you want to apply CSS that may target the TBODY.

So, I don't know how to solve the issue with your test file, but restoring the old AC_TR is probably not the solution. (If it solves it, it's a side effect :)


```diff
+// Note: AC_TD and AC_TR may kill a table nested inside an other table,
+// so such nested tables won't work in a .html file (but will in a .epub
+// that uses the NON-auto-closing ldomDocumentWriter).
```
Means that with `TABLE > TR > TD > TABLE > TR > TD`, the 2nd TR will auto close the first TD, so messing with the nested tables - and I can't imagine the result :)

Frenzie commented:
I wouldn't be too concerned about how crengine used to handle broken HTML before HTML5 defined error handling, but deviations from HTML5/Blink/Gecko/WebKit (if any) are generally undesirable.

Anyway, I just see broken `<dd>` autoclosing? Tables seem perfectly fine at a glance.

Frenzie commented:
But without proper testcases focusing on a single issue it's hard to tell.

poire-z commented:
I also don't mind much broken HTML being not rendered perfectly - and I'm quite convinced there is no solution that works in all cases.
And crengine's single list of parent element to check for - without an additional list of elements to stop previous checks at when met - can't do that well.

But one thing that we should ensure is that the autoclose rules (no matter how limited they are) should not break perfectly balanced and logical simple HTML - which the old AC_TR rule did for the simple table above  (and the new rule still does for nested tables).

virxkane commented:
> I wouldn't be too concerned about how crengine used to handle broken HTML before HTML5 defined error handling

Where is it broken? These tags do not require closing tags.

> Tables seem perfectly fine at a glance.

![Screenshot_20200807_145708](https://user-images.githubusercontent.com/36960933/89639275-50405280-d8be-11ea-948a-4989448f301e.png)

Yeah, perfectly fine. (try to re-render after first render)

How it looks in CoolReader 3.2.38:
![Screenshot_20200807_150259](https://user-images.githubusercontent.com/36960933/89639700-1de32500-d8bf-11ea-8d82-e24b2c9015ee.png)


> But without proper testcases focusing on a single issue it's hard to tell.

Why don't you like this test? Ok, try this file 
[Пайпс_Русская+революция_1.chm.zip](https://github.com/koreader/crengine/files/5040959/_.%2B._1.chm.zip)

pkb commented:
I already don't remember how crengine handles autcolosing. According to html standard:

> A tr element's end tag may be omitted if the tr element is immediately followed by another tr element, or if there is no more content in the parent element.

How does crengine handle there is no more content in the parent element? tr's parents are thead, tbody,tfoot,table?

virxkane commented:
How this file looks in Firefox 
![Screenshot_20200807_151227](https://user-images.githubusercontent.com/36960933/89640365-79fa7900-d8c0-11ea-877b-da0eb5b8bba9.png)

Frenzie commented:
@virxkane 
> Where is it broken? These tags do not require closing tags.

I thought you were talking about a file with errors when you said problematic. If the file doesn't have errors I wouldn't call it problematic. My point was simply that "crengine used to do X" may not be the same thing as "HTML4 said X," let alone "HTML5 says X."

As an aside, `<dd>` is only supposed to occur inside of a `<dl>`.

> Yeah, perfectly fine. (try to re-render after first render)

I see this. There's no change on rerendering.
![Screenshot_2020-08-07_13-43-01](https://user-images.githubusercontent.com/202757/89642338-f0dd4500-d8b3-11ea-96ba-3de82113f43d.png)

> Why don't you like this test? Ok, try this file

I'm talking about isolated test case creation. A big document containing hundreds of pages is the exact opposite. ;-) There are all kinds of distracting variables in your test case, making it hard to tell what's going on. There are at least two separate isolated test cases in there.

virxkane commented:
> As an aside, `<dd>` is only supposed to occur inside of a `<dl>`.

Ok, skipping this.



> I see this. There's no change on rerendering.

Ok, may be it's CoolReader peculiarity, but this table rendering incorrectly: where borders? Why are the table cells jumbled?



> I'm talking about isolated test case creation. A big document containing hundreds of pages is the exact opposite. ;-)

You do not understand me, there is a problem file "Пайпс_Русская+революция_1.chm.zip", which takes a very long time to open and render. Choose "proper testcases focusing on a single issue" from it to your taste, if, of course, you want to help fix this bug.

poire-z commented:
> where borders? Why are the table cells jumbled?

We use our epub.css with HTML/CHM files (while CR old & new use htm.css/chm.css). We obsoleted htm.css and chm.css and don't update them.
epub.css do as browsers do and don't put borders on tables.
I don't understand why you get borders :) you don't have it in your CSSs, and there's no code I can find that handle `<TABLE border>`

Frenzie commented:
@virxkane 

> You do not understand me, there is a problem file "Пайпс_Русская+революция_1.chm.zip", which takes a very long time to open and render. Choose "proper testcases focusing on a single issue" from it to your taste, if, of course, you want to help fix this bug.

Indeed, you've completely lost me. If this document is not related to autoclosing at all why did you bring it up? :-/

Frenzie commented:
> Ok, may be it's CoolReader peculiarity, but this table rendering incorrectly: where borders? Why are the table cells jumbled?

They're jumbled because of the `<dd>` autoclosing issue, which is what's obscuring any potential table issue. Hence at least two separate test cases. :-)

Frenzie commented:
Even if there were no autoclosing issue you wouldn't want that kind of stuff in your test case obscuring things.

poire-z commented:
I guess that initial test case is a good start to refine it to an even smaller one :)  And may be we won't find a smaller one, and it's the whole combination that causes the issue.
(Haven't had the time yet to investigate, but indeed, my gut feeling is that the cause might not be the autoclosing of table elements - but of something else, or bad interactions with what's before)

Frenzie commented:
> I guess that initial test case is a good start to refine it to an even smaller one :)

Absolutely, I intended it as a PSA as much as anything. :+1: 

15 years ago I occasionally reported bugs to Opera, which is when I was given some friendly pointers to improve my test cases. I liked to write minimal HTML, of this variety:

```html
<!DOCTYPE html>
<title>test case</title>
<p style="color:blue">This should be blue but it's red! (Okay, obviously this never happened.)
```

And they told me that generally speaking I should first try to write it like this instead to ensure it's truly an issue about incorrect color parsing:

```html
<!DOCTYPE html>
<html>
<head>
<title>test case</title>
</head>
<body>
<p style="color:blue">This should be blue but it's red! (Okay, obviously this never happened.)</p>
</body>
</html>
```

In other words, an ideal minimal testcase only includes the absolute minimum to reproduce the issue but it avoids going so minimal that it might introduce confounding issues.

The above is a somewhat badly chosen illustration since the "confounding issues" from my sample are the entire point. Perhaps it should be reversed for illustrative purposes instead:

```html
<!DOCTYPE html>
<title>test case</title>
<p style="color:blue">This should render but it doesn't because the DOM thinks there's no BODY element!
(So what's that style=color:blue doing there? It's just distracting. :-) )
```

Btw, @a-kohout is excellent at providing minimal test cases. See https://github.com/koreader/koreader/issues/3192 for example. :-)

Frenzie commented:
If we remove some of the confounding/obscuring factors you still get a plenty weird result.
```html
<!DOCTYPE html>
<html>
<head>
<title>Table autoclosing</title>
</head>
<body>
<table>

<tr>
<td>
<b>Сферы производства</b>

<td>
<b>1890</b>

<td>
<b>1900</b>

<td>
<b>Прирост</b>

<tr>

<td>
Чугун, <i>т</i>

<td>
927 100

<td>
2 933 700

<td>
216%

</table>

</body>
</html>
```

![Screenshot_2020-08-07_15-04-34](https://user-images.githubusercontent.com/202757/89648310-51be4a80-d8bf-11ea-9828-65c65aca6458.png)

![Screenshot_2020-08-07_15-04-56](https://user-images.githubusercontent.com/202757/89648340-5e42a300-d8bf-11ea-948c-f17ac0056f0a.png)

Frenzie commented:
Comparatively, closing manually with a `</tr>` renders as expected:

![Screenshot_2020-08-07_15-08-06](https://user-images.githubusercontent.com/202757/89648665-dad58180-d8bf-11ea-8230-4240cc34be88.png)

poire-z commented:
(Another thing that can be at play : `Tables: complete incomplete tables`  #328 )

virxkane commented:
> Indeed, you've completely lost me. If this document is not related to autoclosing at all why did you bring it up? :-/

Sure, please read again:

> This is a slightly simplified snippet from a real problematic chm file.



> Пайпс_Русская+революция_1.chm.zip

virxkane commented:
I just showed you how changes in one line broke the auto-closing of the <tr> tag.
I did not write here all the way as I came to this, here it is not appropriate. But it seems to me that this is the problem. Maybe I'm wrong, it happens :) But instead of trying to prove to me that I'm wrong, just look at this file "Пайпс_Русская+революция_1.chm.zip".

virxkane commented:
> Comparatively, closing manually with a `</tr>` renders as expected:

Now go ahead and restore the string ```static const char * AC_TR[] = {"tr", "tr", "thead", "tfoot", "tbody", NULL};```.
I not this that its the best solution, I just showed you the problem.

Frenzie commented:
@virxkane I'm not trying to prove you wrong, unless you disagree about effective communication through minimal test cases. In that case I am. ^_^

However, your proposed fix only renders correctly in this particular case somewhat by accident. The resulting DOM is incorrect even if we feed it a fully fledged out table as included below. I would agree it seems like a reasonable workaround since `thead`, `tfoot` and `tbody` are *mostly* somewhat redundant, although if used for styling through CSS that'd fail to work as intended.

```html
<!DOCTYPE html>
<html>
<head>
<title>Complete table</title>
<style>
thead {color:red}
tbody {color:blue}
</style>
</head>
<body>

<table>
<thead>
    <tr>
        <th>Сферы производства</th>
        <th>1890</th>
        <th>1900</th>
        <th>Прирост</th>
    </tr>
</thead>
<tbody>
    <tr>
        <td>Чугун, <i>т</i></td>
        <td>927 100</td>
        <td>2 933 700</td>
        <td>216%</td>
    </tr>
</tbody>
</table>

</body>
</html>
```

The intended rendering looks like:
![Screenshot_2020-08-07_15-57-52](https://user-images.githubusercontent.com/202757/89653072-c34dc700-d8c6-11ea-9dc0-ec56aae37f39.png)

But since the `thead` and `tbody` are treated incorrectly in your suggested patch, it just remains black.
![Screenshot_2020-08-07_15-58-20](https://user-images.githubusercontent.com/202757/89653145-deb8d200-d8c6-11ea-81e8-5527f4fdb9b6.png)

By comparison, here's the correct DOM with the correct rendering as in the current code.
![Screenshot_2020-08-07_15-59-04](https://user-images.githubusercontent.com/202757/89653203-f85a1980-d8c6-11ea-9431-9e54d1ee54dd.png)

I would agree the damage is worse without your suggested workaround, but it's definitely a workaround, not a fix. :-)

virxkane commented:
> So, for this case, I'd say it's a limitation of the AutoClose algorithm, that is stopping at the first parent in the AC_TR list, and won't close the preceeding row TR.

Seem like that, thank you.

> I'm not trying to prove you wrong, unless you disagree about effective communication through minimal test cases. In that case I am. ^_^

Ok. The next sample will be smaller :)


> However, your proposed fix only renders correctly in this particular case somewhat by accident.

No, no, no. I don't proposed this fix, I show problem, I ask how to fix. Point.

poire-z commented:
So, anyway, if none of us want to think too much about that, we could agree on this scenario:

```diff
-static const char * AC_TR[] = {"tr", "tr", "thead", "tfoot", "tbody", NULL};
+static const char * AC_TR[] = {"tr", "tr", "td", NULL};
```
I did this in 2018 because having THEAD, TBODY... is wrong.
I added the `"td"` because it felt like the inverse correct solution, without too much thinking.
But logically, it shouldn't be needed: if there was an opening TD (and it must have one even for dirty HTML publishers that don't close anything - or no content would be shown), it will be closed by pop() when closing the TR. So no need to have in in AC_TR.
So, let's just remove the "td" :)
```diff
-static const char * AC_TR[] = {"tr", "tr", "td", NULL};
+static const char * AC_TR[] = {"tr", "tr", NULL};
```

But after that, we're not yet done with your initial test case :(

poire-z commented:
Same reason for the imbricated DD.
`AC_DD[] = {"dd", "dd", "p", NULL};`
A DD will close a DD - except if there was a P inside the previous DD: it will then close only the P, and will stay inside the previous DD.
While Firefox properly have them not-imbricated and at the same level:

Firefox inspector | Firefox | KOReader | KOReader DOM:
![image](https://user-images.githubusercontent.com/24273478/89663807-6f4ade80-d8d6-11ea-88aa-0de6ee7f902d.png)

poire-z commented:
```html
  <body>
<table border>
<tr>
```
renders well:
<kbd>![image](https://user-images.githubusercontent.com/24273478/89664322-4414bf00-d8d7-11ea-9636-d709e0367540.png)</kbd>

Just add a DD between BODY and TABLE, and boum:
```html
  <body>
<dd>
<table border>
<tr>
```
<kbd>![image](https://user-images.githubusercontent.com/24273478/89664396-6a3a5f00-d8d7-11ea-9b41-2c6bb942544a.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/89664595-aec5fa80-d8d7-11ea-8e3d-82433a40e90f.png)</kbd>

Ok, further tests by removing stuff or adding closing tags to avoid triggering some auto-close:
```html
  <body>
ABC
<dd>
<table border>
<tr>
<td width=148>
<dd><b>Ð¡Ñ~DÐµÑ~@Ñ~K Ð¿Ñ~@Ð¾Ð¸Ð·Ð²Ð¾Ð´Ñ~AÑ~BÐ²Ð°</b></dd>
</td>
<td width=72>
<p align=center><b>1890</b></p>
</td>
</tr>
</table>
  </body>
```
It's the 2nd DD instide the table cell that causes the first one to be auto-closed - even if that 2nd one is properly closed:
<kbd>![image](https://user-images.githubusercontent.com/24273478/89665229-a0c4a980-d8d8-11ea-98b7-20dce4326837.png)</kbd>

So, all that mess is also caused by:
`AC_DD[] = {"dd", "dd", "p", NULL};`
When we meet the 2nd DD in the table cell TD: we go check all the parents until we find a DD or P, and we close all the elements met up to there: the TD, the TR, the TABLE, and the DD.
And what comes next is no more in a TABLE, and just a bunch of block elements.

Remove the DD in the all the table cells, and the TABLE will render well, properly inside that first DD.

I have no idea why it all looked fine in CoolReader 3.2.38...

poire-z commented:
I tried reading [the specs](https://html.spec.whatwg.org/multipage/parsing.html#closing-elements-that-have-implied-end-tags) but I finally [gave up](https://github.com/koreader/koreader/issues/6482#issuecomment-671449935) and went back at the original AutoClose() code and tried to make it do what we expect - and not fail on the problems noticed (nested elements, stopping too early) - may be somehow doing it in the spirits of what I got from the specs.

This seems to solve the issues noticed above.
```diff
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -12952,24 +12952,46 @@ void ldomDocumentWriterFilter::appendStyle( const lChar16 * style )
 void ldomDocumentWriterFilter::AutoClose( lUInt16 tag_id, bool open )
 {
     lUInt16 * rule = _rules[tag_id];
     if ( !rule )
         return;
     if ( open ) {
         ldomElementWriter * found = NULL;
+        ldomElementWriter * found_fallback = NULL;
         ldomElementWriter * p = _currNode;
         while ( p && !found ) {
             lUInt16 id = p->_element->getNodeId();
             for ( int i=0; rule[i]; i++ ) {
                 if ( rule[i]==id ) {
-                    found = p;
-                    break;
+                    if ( i == 0 ) {
+                        // rule[0] should be the same as the element that we see opening
+                        // If we meet one, we're done auto-closing
+                        found = p;
+                        break;
+                    }
+                    else {
+                        // rule[1+] are other elements that we can meet and that we
+                        // should auto-close to become a sibling of them, but only
+                        // when no same-element (from rule[0] has been found.
+                        // Not sure if we should stop at the nearest one met,
+                        // or take the upper most one.
+                        found_fallback = p;
+                    }
                 }
             }
             p = p->_parent;
+            if ( p && p->_element->getNodeId() == el_table ) {
+                // Don't cross TABLE: a new DD inside a TABLE CELL should
+                // not auto-close a upper DD that contains this TABLE.
+                // This feels generic enough, so do it in all cases
+                break;
+            }
+        }
+        if ( !found && found_fallback ) {
+            found = found_fallback;
         }
         // found auto-close target
         if ( found != NULL ) {
             bool done = false;
             while ( !done && _currNode ) {
                 if ( _currNode == found )
                     done = true;
```
<kbd>![image](https://user-images.githubusercontent.com/24273478/89806065-db2b8200-db36-11ea-84cf-c0b9c9a45d23.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/89806269-162db580-db37-11ea-8a43-09bfb7fd8a93.png)</kbd>

Frenzie commented:
Possibly Good Enough™ and better either way. But it's much too warm to think clearly over here. :-)

poire-z commented:
Although it might still not be right.
```
static const char * AC_LI[] = {"li", "li", "p", NULL};
static const char * AC_UL[] = {"ul", "li", "p", NULL};  // ??
static const char * AC_OL[] = {"ol", "li", "p", NULL};  // ??
static const char * AC_DD[] = {"dd", "dd", "p", NULL};
static const char * AC_DL[] = {"dl", "dt", "p", NULL};
static const char * AC_DT[] = {"dt", "dt", "dd", "p", NULL};
```
(these are shifted one slot, so rule[0] is actually the 2nd element

I'm handling `rule[0] should be the same as the element that we see opening` which works for the LI and DD above - but not for UL.
A new UL should never close any upper UL I guess.
Should it close the current LI ? I don't think so, it can be a child UL > LI > UL
Should it close an upper LI ? I don't think so: UL > LI > DIV > P + LI
Should it close stop at the first upper LI ?
So, no rule for LI ?
Or juste remove the 2nd "lu" and let It close and stop at any upper P ?

poire-z commented:
Pfffffff. If we update the AC_* lists or the AutoClose() algorithm, we might have a totally different DOM, and previously made bookmarks/highlights would not work.
And our normalized XPointers and their migration only deal with the added "internal boxing" elements - so it's impossible to do any migration if we change the auto closing stuff as some elements will have new parents (which is possible even on documents that did render quite well and had no noticable issue).

So, that would mean a new DOM_VERSION - and keep using the old AC_ and AutoClose() function for older DOM versions, and switch to new ones for new books - with no migration.
With KOReader, it would be the default behaviour (to not upgrade) - but I think CoolReader always upgrade to the latest DOM version, right ?
And each time we'll fix that autoclose stuff because of a bad decision made now or some unforeseen case, we'll need to do that again.

If I trust some past comments of mine:
- ldomDocumentWriterFilter and the AutoClose stuff is only used with **HTML, CHM and PDB(html)**.
- EPUB, FB2, RTF, WORD, plain text, PDB(txt) use ldomDocumentWriter which doesn't do AutoClose.

So, should we care about not breaking past highlights for these formats that I consider little-used document formats ?
Or can we just ignore this issue ? :/ :)

Frenzie commented:
I'll drop these tests here https://github.com/html5lib/html5lib-tests/blob/71eebd59772d1d39aced0c0582ae9c09acf3ce6e/serializer/optionaltags.test

My view is similar to yours. Breaking it should be relatively low-impact. But if we do, unfortunately that makes the option of incremental improvements (i.e., continuous breakage) worse than it might have been.

poire-z commented:
So, I guess my first idea [above](https://github.com/koreader/crengine/pull/243#issuecomment-671457077) is not right (assuming the second element is the same as the element and we should stop there).
I came up with this:
```diff
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -12954,21 +12954,50 @@ void ldomDocumentWriterFilter::AutoClose( lUInt16 tag_id, bool open )
     lUInt16 * rule = _rules[tag_id];
     if ( !rule )
         return;
-    if ( open ) {
+    if ( open ) { // called by OnTagOpen()
         ldomElementWriter * found = NULL;
         ldomElementWriter * p = _currNode;
-        while ( p && !found ) {
-            lUInt16 id = p->_element->getNodeId();
+        lUInt16 id = p ? p->_element->getNodeId() : 0;
+        while ( p ) {
             for ( int i=0; rule[i]; i++ ) {
                 if ( rule[i]==id ) {
                     found = p;
-                    break;
+                    // We don't stop at the first one met: we continue checking
+                    // upper nodes that this node might close. But we might break
+                    // below when meeting one of the stop element.
                 }
             }
             p = p->_parent;
+            if ( !p )
+                break;
+            id = p->_element->getNodeId();
+            // Some hardcoded bits (to not complicate AC_* rules and their parsing...)
+            // for parent elements that should stop us.
+            if ( id == el_table ) {
+                // Never cross a TABLE (a new DD inside a TABLE CELL should
+                // not auto-close a upper DD that contains this TABLE).
+                // This feels generic enough, as a table has independant block
+                // rendering contexts for each of its cells.
+                break;
+            }
+            if ( id == el_ruby ) {
+                // RUBY should mostly containe ruby sub-element (and possible
+                // some other inline elements that don't auto close).
+                // So, we shouldn't need to check for tag_id: stop there
+                break;
+            }
+            if ( tag_id == el_li && (id == el_ul || id == el_ol) ) {
+                // A LI stops at its UL or OL container, so it does not
+                // close a upper LI
+                break;
+            }
+            if ( id == el_dl && (tag_id == el_dl || tag_id == el_dt || tag_id == el_dd ) ) {
+                // A DL, DT or DD stop at its DL container, so it does not close it
+                // not any upper DT or DD
+                break;
+            }
         }
-        // found auto-close target
-        if ( found != NULL ) {
+        if ( found != NULL ) { // found auto-close target
             bool done = false;
             while ( !done && _currNode ) {
                 if ( _currNode == found )
```
And this AC_ stuff I re-ordered and commented (with a few changes but not showing the diff as it would make it less readable):
```c
// Most block container elements close a P.
// Most should not close themselves, as they can be nested
// When closing upper nodes, we stop at TABLE to not close them, as their content
// is an independant rendering context (this is hardcoded in AutoClose())
static const char * AC_P[]  = {"p", "p", NULL}; // a P close any P, so there can't be nested <P>s
static const char * AC_DIV[] = {"div", "p", NULL};

// (UL/OL used to have "li" there, but this was wrong as a UL/OL can be inside a LI, with nested lists)
static const char * AC_UL[] = {"ul", "p", NULL};
static const char * AC_OL[] = {"ol", "p", NULL};
// A LI should close a previous LI, but should not close an upper UL or OL, and should not
// go look upper them and close any upper LI (this is hardcoded in AutoClose())
static const char * AC_LI[] = {"li", "li", "p", NULL};

// Description list (DL = list, DT = term, DD = definition/description)
// None of these should close an upper DL containing them, and should not go look upper
// their containing DL and close any upper DD/DT (this is hardcoded in AutoClose())
// A new DL can happen in a DD, but not in a DT.
static const char * AC_DL[] = {"dl", "dt", "p", NULL};
// DT and DD close any previous term or definition
static const char * AC_DT[] = {"dt", "dt", "dd", "p", NULL};
static const char * AC_DD[] = {"dd", "dd", "dt", "p", NULL}; // (used to not close DT, but no reason not to)

// A TABLE should not close anything (except any previous P), because of nested tables
static const char * AC_TABLE[] = {"table", "p", NULL};
// Table rows container should close any previous rows container or standalone rows
static const char * AC_THEAD[] = {"thead", "tr", "thead", "tfoot", "tbody", NULL};
static const char * AC_TBODY[] = {"tbody", "tr", "thead", "tfoot", "tbody", NULL};
static const char * AC_TFOOT[] = {"tfoot", "tr", "thead", "tfoot", "tbody", NULL};
// A TR should close any previous TR (or unclosed previous TD, that would anyway
// be closed when closing its containing TR)
static const char * AC_TR[] = {"tr", "tr", "td", NULL};
// TH and TD close any previous TH or TD
static const char * AC_TH[] = {"th", "th", "td", NULL};
static const char * AC_TD[] = {"td", "td", "th", NULL};

// Ruby elements (often without a closing element and expecting auto closing
// according to various rules, sometimes contradictory...)
// When closing upper nodes, we stop at RUBY to not close upper ruby sub-elements
// and allow nested ruby (this is hardcoded in AutoClose())
// A new RBC and RTC close every other ruby sub-element
static const char * AC_RBC[] = {"rbc", "rbc", "rb", "rtc", "rt", NULL};
static const char * AC_RTC[] = {"rtc", "rtc", "rt", "rbc", "rb", NULL};
// A new RB closes every other ruby sub-element, except RBC (which may contain multiple RB)
static const char * AC_RB[] = {"rb", "rb", "rt", "rtc", NULL};
// A new RT closes every other ruby sub-element, except RTC (which may contain multiple RT)
static const char * AC_RT[] = {"rt", "rt", "rb", "rbc", NULL};
// A RP (ruby fallback parenthesis) closes ruby sub-element, except containers
static const char * AC_RP[] = {"rp", "rp", "rb", "rt", NULL};

// A new one of these should close any previous same one
static const char * AC_OPTION[] = {"option", "option", NULL};
static const char * AC_PARAM[] = {"param", "param", NULL};
static const char * AC_PRE[] = {"pre", "pre", NULL};

// Auto-closing (no sub-element accepted)
static const char * AC_HR[] = {"hr", NULL};
static const char * AC_BR[] = {"br", NULL};
static const char * AC_WBR[] = {"wbr", NULL};
static const char * AC_IMG[]= {"img", NULL};
static const char * AC_COL[] = {"col", NULL};
static const char * AC_BASE[] = {"base", NULL};
static const char * AC_LINK[] = {"link", NULL};
static const char * AC_META[] = {"meta", NULL};
static const char * AC_AREA[] = {"area", NULL};
static const char * AC_EMBED[] = {"embed", NULL};
static const char * AC_INPUT[] = {"input", NULL};
static const char * AC_SOURCE[] = {"source", NULL};
static const char * AC_TRACK[] = {"track", NULL};
```

Not much change to the original AC (except UL/OL not closing any LI, and DT also closing a DD, for symetry), I've just added comment to explain why these rules with what I know about these tags, and not by following the specs. The new stuff is the stop rules in the C code.
This too makes the CHM and the test autoclose HTML fine.

poire-z commented:
I'm still trying to understand the specs, and I need a bit of spec reading wits :) 

There is list of tags that are considered "special", which looks mostly like all elements that are not inline (so, block and others found in head, etc...)
https://html.spec.whatwg.org/multipage/parsing.html#special

Can you look at these steps in the section https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody:
`A start tag whose tag name is "li"`
and the following
`Any other end tag`

They both are bout walking up the parents to close the matching start tag, but they mention:
For start LI: _If node is in the special category, but is not an address, div, or p element, then jump to the step labeled done below._
For any other end tag: _Otherwise, if node is in the special category, then this is a parse error; ignore the token, and return._

Do you understand (like I do) that all these "special" tags are stop tags when walking up parents? (so, with `<span> <ul> <li> </span>` , or with `<dl> <dd> <ul> <li> </dd>`  the last one won't close the first one ? Does that make sense ?

Above the `Any other end tag`, there is:
`An end tag whose tag name is one of: "address", "article", "aside", "blockquote", "button", "center", "details", "dialog", "dir", "div", "dl", "fieldset", "figcaption", "figure", "footer", "header", "hgroup", "listing", "main", "menu", "nav", "ol", "pre", "section", "summary", "ul"`
These would then not have that limitation to stop at a special tag.
Does that make sense?

There's another set:
`A start tag whose tag name is one of: "address", "article", "aside", "blockquote", "center", "details", "dialog", "dir", "div", "dl", "fieldset", "figcaption", "figure", "footer", "header", "hgroup", "main", "menu", "nav", "ol", "p", "section", "summary", "ul"`
plus a few more specific ones, that all should close any previously still open P.

So, it looks like we could do with a few sets of elements, and if we order them correctly in fb2def.h, we could hardcode a few `if (tag_id >= el_special_start && tag_id <= el_special_end)` and do the closing instead of having tons of AC_ that all just close a P.

(There's lots of stuff we won't be able to do easily, like active formatting elements  `a, b, big, code, em, font, i, nobr, s, small, strike, strong, tt, and u`, that when upper and closed inner by some mis-nested elements, should be resurrected and still applied to next content... or mis-nested elements adoptions agency stuff, and some other stacks, and specific modes for when parsing tables... but well :)

Frenzie commented:
> Do you understand (like I do) that all these "special" tags are stop tags when walking up parents? (so, with `<span> <ul> <li> </span>` , or with `<dl> <dd> <ul> <li> </dd>` the last one won't close the first one ? Does that make sense ?

I think so. I think that should be `<span><ul><li></li></ul></span>` and `<dl><dd><ul><li></li></ul></dd></dl>`.

poire-z commented:
Still trying to understant the specs...
A little thing saddened me:
<kbd>![image](https://user-images.githubusercontent.com/24273478/90308716-197fc300-dee2-11ea-982f-b9e16f47ef5a.png)</kbd> <kbd>![image](https://user-images.githubusercontent.com/24273478/90308851-35379900-dee3-11ea-9bba-86ab37e12f41.png)</kbd> 

Everything block closes a P.
So, that would mean that inside a P, you can't have any fancy embedded (as I called them) float with some stuctural block elements (like DIV, HR, TABLE), but only plain inline stuff (like IMG, dropcaps in SPAN...).
So, to do anything fancy, you'd have to close the P, insert the fancy float as a block element, and re-open the P (the algo just closes it, it won't re-open it) - and so, have a line break even if you didn't want one.

Does that sound right? And if yes, what would be the alternative if I wanted a fancy float in the middle of some text, and I didn't want any break in my text? (I see: just not use a P - but then it breaks all the "semantic" idea of using P...) (or is it that in the "semantic" realm, it's a sin to insert some fancy foreign stuff in a P unit that should read without any foreign break?)

poire-z commented:
And nested (misnested or not) Hx tags only close the direct previous - but not if there's an intermediate in-between...:
<kbd>![image](https://user-images.githubusercontent.com/24273478/90309049-0c180800-dee5-11ea-863a-6f6defee418e.png)</kbd>

Frenzie commented:
>  (or is it that in the "semantic" realm, it's a sin to insert some fancy foreign stuff in a P unit that should read without any foreign break?)

As long as it's related (like images or whatever), that should be fine. But indeed they'll be auto-closed by other block-level elements.

poire-z commented:
I'm still working on that HMTL parser. I think I made sense of the specs...
A good part of the stuff like "close emplied end tags" in the algo is just to notice the HTML is bad and flag a "parser error" that the algo just ignore and go do normal work as if nothing was wrong (often just skipping these tags while closing up to the target node).
So, if we don't care about noticing bad HTML and just take the normal behaviour, it's simplified quite a bit.
Still some stuff I have to think about before proposing something, I'd rather get it right as it will need a new DOM_VERSION (I can keep the current behaviour for older dom versions, the new stuff plugs in rather nicely as an alternate algorithm).
Mentionning that again: with KOReader, we won't update the dom_version of previously opened books - but I think you do that with CoolReader. Might be not much of an issue with this one, as it will only affect plain HTML and CHM books.

And there's the [lib.ru hacks](https://github.com/koreader/koreader/issues/6480#issuecomment-676128803) in that parser to possibly try to solve.

(All that specs reading and work for so little personal benefit :), as most formats like FB2 or EPUB don't use that HTML parser... but better get something right than add more hacks :/ )

virxkane commented:
> A good part of the stuff like "close emplied end tags" in the algo is just to notice the HTML is bad and flag a "parser error" that the algo just ignore and go do normal work as if nothing was wrong (often just skipping these tags while closing up to the target node).

Good news.


> Mentionning that again: with KOReader, we won't update the dom_version of previously opened books - but I think you do that with CoolReader.

For me, this is still an open question. Now, in the program settings, you we choose 'legacy' or 'latest'. And this choice is saved for every book. If we remove this option, we will have to do something with the bookmarks saved in the database.
